### PR TITLE
new script to sync up the repo with official helm repo

### PR DESCRIPTION
**Pull request by hasancelik**
With this PR;

- created `create-pr-at-official-helm-repo` script which is used by [helm-chart-hazelcast-sync](http://jenkins.hazelcast.com/view/Plugins/job/helm-chart-hazelcast-sync/) and [helm-chart-hazelcast-jet-sync](http://jenkins.hazelcast.com/view/Plugins/job/helm-chart-hazelcast-jet-sync/)
jenkins jobs. These jobs are triggered when someone pushes changes to master which affects `/stable/hazelcast` or `stable/hazelcast-jet` folders:
Screen Shot 2019-09-05 at 18 01 47

These two jobs are using github tokens which are bound to my github account and `devOpsHazelcast`  therefore we need to add them as an approver/reviewer to `helm/charts` repo. Modifying `OWNERS` file is not enough, I guess. Do you have any idea @leszko ?

- as you can see at `create-pr-at-official-helm-repo` script, hazelcast and hazelcast-jet folders are copied into local `helm/charts` so this repo and official repo start to use same `README` file. To change its content based on repo, I need to inject some logic into `README` files so I have converted them into asciidoc:

```
:repo: hazelcast/charts
...
ifeval::["{repo}"=="hazelcast/charts"]
[source,bash]
----
$ helm install --name my-release -f values.yaml hazelcast/hazelcast-jet
----
endif::[]

ifeval::["{repo}"=="helm/charts"]
[source,bash]
----
$ helm install --name my-release -f values.yaml stable/hazelcast-jet
----
endif::[]
```

Github support `.adoc` README files but I am not sure `helm/charts` repo owners allow it. We will learn when I create PR at official repo or we can ask them directly via slack or github issue. cc: @leszko

### CODE REVIEW

File: create-pr-at-official-helm-repo [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+CHARTS=(hazelcast hazelcast-jet);
+if [[ " "${CHARTS[@]}" " == *" "$1" "* ]] ;then
+    echo "$1: chart is valid"
+else
+    echo "$1: not recognized. Valid charts are:"
+    echo "${CHARTS[@]/%/,}"
+    exit 1
+fi
+CHART_NAME=$1
+
```
Comment (leszko) [inline create-pr-at-official-helm-repo]: Consider checking (and maybe even installing) dependencies like `git`, `hub`, etc.

An example how we do it in other scripts: https://github.com/hazelcast/hazelcast-qe/blob/master/it/ocp-automation/publish.sh#L63

Comment (hasancelik) [reply create-pr-at-official-helm-repo]: Related jenkins jobs are using [hub-jenkins-ci](https://hub.docker.com/r/hazelcast/hub-jenkins-ci) docker image so I think we do not need to check dependencies are installed or not.

**Comment by hasancelik** [comment]
> We should not approve form devOpsHazelcast account, at least not start from that. I'd keep the first version of this script with the manual approval. Just mention every OWNER in the PR description, like add Please approve @leszko @mesutcelik @googlielmo @eminn. Better to do it step by step imo.

it seems good approach. I will update comment part 👍 

> Do you signoff the commits? (I think that's the requirement in the official Helm Chart repo).

yes:
```
...
git config --global user.email "${GITHUB_EMAIL}"
git config --global user.name "${GITHUB_USERNAME}"
...
git commit -s -m "[sync] ${LATEST_COMMIT_MESSAGE}"
```
> I'm not sure if we shouldn't be more smart with this README. This ifeval is smart, but I'm not sure if that's the right approach. Did you see anything like that in any chart? Maybe another option would be to keep kind-of template in our repo and generate the proper README.md from that? Or some other thing?

If the setup is not smart enough, we need to update official README manually for every commit affects README. `ifeval/asciidoc` approach is the only solution I have found. There is no chart using `.adoc` README but as i said before asciidoc READMEs are supported by Github and I did not find any limitation/rule about it at official helm chart repo docs.

> I'd not include [sync] in the commit message, since I'm not sure they'll like it. You open the official repo and the first thing you see are some sync commits ;)

I have added it to make easy to follow automated PRs at official repo but you are right it can draw attention 🙂 

> Maybe a stupid question, but did you try this this script on your fork of the official Helm Chart repo? :)

I did, here is the test PRs 🙂 :
https://github.com/hasancelik/charts-1/pull/4

**Comment by hasancelik** [comment]
With https://github.com/hazelcast/charts/pull/57/commits/2d38296d789f9ab6363084c13048123035c77f45 commit, the script started use restricted `devOpsHelm` user. Thanks to @jerrinot for advice.

**Comment by hasancelik** [comment]
@leszko PR rebased. Comments are addressed and answered. Could you review again?

**Comment by hasancelik** [comment]
verify

**Comment by hasancelik** [comment]
verify

**Comment by hasancelik** [comment]
verify

**Comment by hasancelik** [comment]
verify

File: create-pr-at-official-helm-repo [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    git remote -v
+    git add ./stable/${CHART_NAME} && git commit -s -m "${LATEST_COMMIT_MESSAGE}"
+    git push -u devops-helm ${BRANCH_NAME}
+
+    OFFICIAL_HELM_PR_ISSUE_NUMBER=$(hub pull-request -m "[stable/${CHART_NAME}] ${LATEST_COMMIT_MESSAGE}" -b ${HELM_REPOSITORY}:master -h devOpsHelm/charts:${BRANCH_NAME} | awk -F/ '{print $7}')
+    # hub api repos/${HELM_REPOSITORY}/issues/${OFFICIAL_HELM_PR_ISSUE_NUMBER}/comments -f body='/ok-to-test'
+    # hub api repos/${HELM_REPOSITORY}/issues/${OFFICIAL_HELM_PR_ISSUE_NUMBER}/comments -f body='/lgtm'
+    hub api repos/${HELM_REPOSITORY}/issues/${OFFICIAL_HELM_PR_ISSUE_NUMBER}/comments -f body='Please approve @hasancelik @leszko @mesutcelik @googlielmo @eminn'
+
+    # export GITHUB_TOKEN=${APPROVER_GITHUB_TOKEN}
```
Comment (leszko) [inline create-pr-at-official-helm-repo]: The same, do not leave commented-out code. If you need something for later, just create a GH Issue.

File: create-pr-at-official-helm-repo [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (leszko) [inline create-pr-at-official-helm-repo]: Add a check for the tools you use in this script (`ssh`, `hub`, `git`) or at least mention them in the comment. It's a good practice, then you let know the user of the script what is needed (fail-fast). I know that it's now run on Jenkins, so there is no problem with missing dependencies, but still, someone may look at this script in the future and need it.

**Comment by hasancelik** [comment]
> @hasancelik
> 
> Added some minor comments. One more thing that comes to my mind is that normally all PRs should be synced. But what if there is a PR that does not change anything in `stable/hazelcast` (neither in `stable/hazelcast-jet`). Will it still send the PR to the Official Helm Chart repo?

As I mentioned at first comment, jenkins jobs are triggered **only** by modifications on `stable/hazelcast` and `stable/hazelcast-jet, so it will not send the PR.

File: create-pr-at-official-helm-repo (hunk 2 above)
Comment (hasancelik) [reply create-pr-at-official-helm-repo]: 👍 for future reference --> https://github.com/hazelcast/charts/issues/65

File: create-pr-at-official-helm-repo [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    sed -i 's#:repo:.*#:repo: helm/charts#g' ./stable/${CHART_NAME}/README.adoc
+    git checkout -b ${BRANCH_NAME}
+    git remote add devops-helm git@github.com:devOpsHelm/charts.git
+    git remote -v
+    git add ./stable/${CHART_NAME} && git commit -s -m "${LATEST_COMMIT_MESSAGE}"
+    git push -u devops-helm ${BRANCH_NAME}
+
+    OFFICIAL_HELM_PR_ISSUE_NUMBER=$(hub pull-request -m "[stable/${CHART_NAME}] ${LATEST_COMMIT_MESSAGE}" -b ${HELM_REPOSITORY}:master -h devOpsHelm/charts:${BRANCH_NAME} | awk -F/ '{print $7}')
+    # hub api repos/${HELM_REPOSITORY}/issues/${OFFICIAL_HELM_PR_ISSUE_NUMBER}/comments -f body='/ok-to-test'
+    # hub api repos/${HELM_REPOSITORY}/issues/${OFFICIAL_HELM_PR_ISSUE_NUMBER}/comments -f body='/lgtm'
```
[the remark this answers is not in the document]
Comment (hasancelik) [reply create-pr-at-official-helm-repo]: 👍

File: create-pr-at-official-helm-repo (hunk 3 above)
Comment (hasancelik) [reply create-pr-at-official-helm-repo]: You are right, it would be good for the future 👍

**Comment by hasancelik** [comment]
@leszko Thanks for the review 🙂 All your comments are addressed with latest commits so PTAL 👍

**Pull request merged**